### ISSUE TITLE
Working with ApiAuthorizationDbContext using EFCore extensions and Worker Service.

### ISSUE THREAD

**Issue by MARKUSER** [jasontaylordev/CleanArchitecture#379]
1. I have created application using this template. Now i would like to use EFCore extensions (example EFCore.BulkExtensions) to bulk save records. Since we don't have regular DbContext available in ApplicationContext class may i know if there is a possibility to invoke BulkInsert of BulkExtensions framework through ApiAuthorizationDbContext?

public class ApplicationDbContext : ApiAuthorizationDbContext<ApplicationUser>, IApplicationDbContext
    {
        private readonly ICurrentUserService _currentUserService;
        private readonly IDateTime _dateTime;
        private readonly IDomainEventService _domainEventService;

        public ApplicationDbContext(
            DbContextOptions options,
            IOptions<OperationalStoreOptions> operationalStoreOptions,
            ICurrentUserService currentUserService,
            IDomainEventService domainEventService,
            IDateTime dateTime) : base(options, operationalStoreOptions)
        {
            _currentUserService = currentUserService;
            _domainEventService = domainEventService;
            _dateTime = dateTime;
        }

**//I wanted to create a method like this**
 public Task BulkInsert(List of entities)
        {
            
            await  db.BulkInsertAsync(entitiesList);
        }
}

Using regular DbContext, i can call like this _dbContext.BulkInsert(with parameters.)

2. I am not sure if this template can be used with WorkerService project (window service) ? if yes, then i think i shouldn't call  behavior like Authorization in the pipeline and then am not sure if this **ApiAuthorizationDbContext**<ApplicationUser> will be helpful. Auditable entities gets user details from HttpContextAccessor so i wont be able to get these from worker service hosted like windows service.

3. Let say i have some reusable logic which would be commonly used in different Handlers / Queries. May i know where to create this reusable logic ? Can i write this in Services folder of infrastructure ?

I am not sure if i should raise every point as separate ticket. If yes, please let me know.

**Comment by jasontaylordev** [comment]
Firstly, let me apologise for taking so long to respond. The last 3 years have been a very busy time and I haven't been able to attend to this project as much as I would like. Secondly, I realise that you probably have already found solutions to these problems, but in case you have not, and at least for the benefit of future readers, I'll share some now.

> 1. You would like to utilise EF Core extensions (such as [EFCore.BulkExtensions](https://github.com/borisdj/EFCore.BulkExtensions)).

Add the **EFCore.BulkExtensions** package to **Infrastructure**.

Add a new `BulkInsertAsync` method to `ApplicationDbContext`:

```csharp
    public async Task BulkInsertAsync<T>(IList<T> entities)
        where T : class
    {
        await BulkInsertAsync(entities);
    }
```

Within **Application**, update `IApplicationDbContext` to expose the new method:

```diff
public interface IApplicationDbContext
{
    DbSet<TodoList> TodoLists { get; }

    DbSet<TodoItem> TodoItems { get; }

+    Task BulkInsertAsync<T>(IList<T> entities);

    Task<int> SaveChangesAsync(CancellationToken cancellationToken);
}
```
You can now bulk insert entities from the **Application** layer.

> 2. Background services and Windows services.

These services should not access anything but the **Application** layer. The services will use MediatR to invoke relevant commands and queries to perform necessary operations. The services can use dependency injection to provide a valid instance of `ICurrentUserService`. The instance will have a simple implementation and likely represent some SYSTEM user.

Remember that while the current implementation of `ICurrentUserService` depends on `IHttpContextAccessor`, the auditable entities functionality only depends on `ICurrentUserService`. You can switch the `ICurrentUserService` implementation to something more suitable for background services.

> 3. Duplication of logic across request handlers.

Jimmy Bogard covers this very will in his blog post [Dealing with Duplication in MediatR Handlers](https://lostechies.com/jimmybogard/2016/12/12/dealing-with-duplication-in-mediatr-handlers/) and his talk [Domain-Driven Refactoring](https://youtu.be/_dQRAsVhCqA).

I hope someone finds this helpful. Happy Coding! 😀

**Comment by MARKUSER** [comment]
Thanks @jasontaylordev for taking time and explaining. I have implemented #1 and #2 but certainly i will take a look into blogs you have posted.

**Comment by KevinDeBeer** [comment]
@MARKUSER Could you provide an example of how you created a BackgroundService?
I've been trying to properly create one for a while now, but for some reason I always get entangled in DI problems..
